File: todo/test_gmt_fft_2d.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <complex.h>
#include "gmt.h"

int main() {
    void *API;
    unsigned int n_columns = 4;
    unsigned int n_rows = 4;
    int direction = GMT_FFT_FWD; // Forward transform
    float complex data[] = {1 + 0*I, 2 + 0*I, 3 + 0*I, 4 + 0*I,
                            5 + 0*I, 6 + 0*I, 7 + 0*I, 8 + 0*I,
                            9 + 0*I, 10 + 0*I, 11 + 0*I, 12 + 0*I,
                            13 + 0*I, 14 + 0*I, 15 + 0*I, 16 + 0*I};

    // Initialize the GMT API
    if ((API = GMT_Create_Session("test_gmt_fft_2d", 2, 0, NULL)) == NULL) {
        printf("Error: Could not initialize the GMT API.\n");
        return EXIT_FAILURE;
    }

    // Call the GMT_FFT_2D function
    if (GMT_FFT_2D(API, (float *)data, n_columns, n_rows, direction, GMT_FFT_COMPLEX) != GMT_NOERROR) {
        printf("Error: Could not perform the 2D FFT.\n");
        return EXIT_FAILURE;
    }
                                                                         
    // Print the transformed data                                        
    for (int i = 0; i < n_rows; ++i) {                                   
        for (int j = 0; j < n_columns; ++j) {                            
            printf("(%f, %f) ", crealf(data[i * n_columns + j]), cimagf(data[i * n_columns + j]));
        }                                                                
        printf("\n");                                                    
    }                                                                      

    // Clean up the GMT API session
    GMT_Destroy_Session(API);

    return EXIT_SUCCESS;
}

File: todo/test_gmt_fft_2d.py
import numpy as np

def main():
    n_columns = 4
    n_rows = 4
    
    data = np.array([
        [1 + 0j, 2 + 0j, 3 + 0j, 4 + 0j],
        [5 + 0j, 6 + 0j, 7 + 0j, 8 + 0j],
        [9 + 0j, 10 + 0j, 11 + 0j, 12 + 0j],
        [13 + 0j, 14 + 0j, 15 + 0j, 16 + 0j]
    ], dtype=np.complex64)
    
    # Perform the 2D FFT
    transformed_data = np.fft.fft2(data)
    
    # Print the transformed data
    for i in range(n_rows):
        for j in range(n_columns):
            print("({:.1f}, {:.1f}) ".format(transformed_data[i, j].real, transformed_data[i, j].imag), end="")
        print()
        
if __name__ == "__main__":
    main()
